### PR TITLE
Fix all the memory issues

### PR DESCRIPTION
**Pull request by jay18001**
1e80o8

### CODE REVIEW

File: Example/Source/Navigation/NavigationDetailViewController.swift [hunk 1]
```diff
@@ -19,7 +19,7 @@ class NavigationDetailViewController: BrickViewController {
     var showSubItems: Bool = false
 
     // Mark: - Overrides
-
+
```
Comment (klundberg) [inline Example/Source/Navigation/NavigationDetailViewController.swift]: Can these whitespace changes be reverted?

**Comment by jay18001** [comment]
Fixed the failing test cases

File: Source/Bricks/Button/ButtonBrick.swift [hunk 2]
```diff
@@ -57,9 +57,12 @@ public class ButtonBrick: Brick {
...
 
+    private var model: ButtonBrickCellModel?
+
     convenience public init(_ identifier: String = "", width: BrickDimension = .Ratio(ratio: 1), height: BrickDimension = .Auto(estimate: .Fixed(size: 50)), backgroundColor: UIColor = UIColor.clearColor(), backgroundView: UIView? = nil, title: String, configureButtonBlock: ConfigureButtonBlock? = nil, onButtonTappedHandler: ButtonTappedBlock? = nil) {
         let model = ButtonBrickCellModel(title: title, configureButtonBlock: configureButtonBlock, onButtonTappedHandler: onButtonTappedHandler)
         self.init(identifier, width: width, height: height, backgroundColor: backgroundColor, backgroundView: backgroundView, dataSource: model, delegate: model)
+        self.model = model
     }
 
     public init(_ identifier: String = "", width: BrickDimension = .Ratio(ratio: 1), height: BrickDimension = .Auto(estimate: .Fixed(size: 50)), backgroundColor: UIColor = .clearColor(), backgroundView: UIView? = nil, dataSource: ButtonBrickCellDataSource, delegate: ButtonBrickCellDelegate? = nil) {
```
Comment (rubencagnie) [inline Source/Bricks/Button/ButtonBrick.swift]: Maybe the assignment of the model be in here, so if someone initializes the `ButtonBrick` with a `ButtonBrickCellModel` it also works (like you did for the `CollectionBrick`)

File: Source/Bricks/Collection/CollectionBrick.swift [hunk 3]
```diff
@@ -169,19 +180,20 @@ public class CollectionBrickCell: BrickCell, Bricklike, AsynchronousResizableCel
         brickCollectionView.layout.delegate = self
 
         brickCollectionView.beginConfiguration {
-            self.brick.dataSource.configure(for: self)
+            self.brick.dataSource?.configure(for: self)
```
Comment (rubencagnie) [inline Source/Bricks/Collection/CollectionBrick.swift]: DataSource is already unwrapped, so I would propose to use that one

File: Source/Bricks/Collection/CollectionBrick.swift [hunk 4]
```diff
@@ -169,19 +180,20 @@ public class CollectionBrickCell: BrickCell, Bricklike, AsynchronousResizableCel
...
         }
         
         brick.brickTypes.forEach {
             self.brickCollectionView.registerBrickClass($0)
         }
         
-        brick.dataSource.registerBricks(for: self)
+        brick.dataSource?.registerBricks(for: self)
```
Comment (rubencagnie) [inline Source/Bricks/Collection/CollectionBrick.swift]: DataSource is already unwrapped, so I would propose to use that one

File: Source/Bricks/Collection/CollectionBrick.swift [hunk 5]
```diff
@@ -169,19 +180,20 @@ public class CollectionBrickCell: BrickCell, Bricklike, AsynchronousResizableCel
...
         
         brickCollectionView.collectionInfo = CollectionInfo(index: self.index, identifier: self.brick.identifier)
         
-        let section = brick.dataSource.sectionForCollectionBrickCell(self)
-        brickCollectionView.setSection(section)
+        if let section = brick.dataSource?.sectionForCollectionBrickCell(self) {
```
Comment (rubencagnie) [inline Source/Bricks/Collection/CollectionBrick.swift]: DataSource is already unwrapped, so I would propose to use that one

File: Source/Bricks/Image/ImageBrick.swift [hunk 6]
```diff
@@ -9,18 +9,37 @@
...
+        
+        if dataSource is ImageBrickModel || dataSource is ImageURLBrickModel {
+            self.model = dataSource
+        }
+    }
+    
+    public convenience init(_ identifier: String = "", width: BrickDimension = .Ratio(ratio: 1), height: BrickDimension = .Auto(estimate: .Fixed(size: 50)), backgroundColor: UIColor = .clearColor(), backgroundView: UIView? = nil, image: UIImage, contentMode: UIViewContentMode) {
+        let model = ImageBrickModel(image: image, contentMode: contentMode)
+        self.init(identifier, width: width, height: height, backgroundColor:backgroundColor, backgroundView:backgroundView, dataSource: model)
+        self.model = model
```
Comment (rubencagnie) [inline Source/Bricks/Image/ImageBrick.swift]: Is this needed if you have assigned in the default initializer?

File: Source/Bricks/Label/LabelBrick.swift [hunk 7]
```diff
@@ -59,9 +59,14 @@ public class LabelBrick: Brick {
...
     }
-
+    
+    private var model: LabelBrickCellModel?
+    
     convenience public init(_ identifier: String = "", width: BrickDimension = .Ratio(ratio: 1), height: BrickDimension = .Auto(estimate: .Fixed(size: 50)), backgroundColor: UIColor = UIColor.clearColor(), backgroundView: UIView? = nil, text: String, configureCellBlock: ConfigureLabelBlock? = nil) {
-        self.init(identifier, width: width, height: height, backgroundColor: backgroundColor, backgroundView: backgroundView, dataSource: LabelBrickCellModel(text: text, configureCellBlock: configureCellBlock))
+        let model = LabelBrickCellModel(text: text, configureCellBlock: configureCellBlock)
+        self.init(identifier, width: width, height: height, backgroundColor: backgroundColor, backgroundView: backgroundView, dataSource: model)
+        self.model = model
```
Comment (rubencagnie) [inline Source/Bricks/Label/LabelBrick.swift]: Move this to the standard initializer (like CollectionBrick)

File: Tests/ViewControllers/TestBrickViewController.swift [hunk 8]
```diff
@@ -12,9 +12,15 @@ import BrickKit
 class TestBrickViewController: BrickViewController, BrickRegistrationDataSource {
 
     var brickRegistered = false
+    var labelTest = false
```
Comment (rubencagnie) [inline Tests/ViewControllers/TestBrickViewController.swift]: What is this for?

**Pull request merged**